test/python/model_validation/validation_tool.py

```diff
@@ -0,0 +1,184 @@
+import onnxruntime_genai as og
+import argparse
+import time
+from ...src.python.py.models.builder import create_model
+from datasets import load_dataset 
+import torch
+import json
+import tqdm
+ 
+# def calculate_perplexity(args, output_dir):
+#     print("We are now calculating perplexity")
+ 
+#     # actual model object
+#     model = og.Model(f'{output_dir}')
+
+#     # tokenizer should already be created
+#     tokenizer = og.Tokenizer(model)
+#     print(tokenizer)
+
+#     test = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
+
+#     print(test)
+#     # Calculate the perplexity, understand the output of the models
+#     # Understand the output from a model and logits.
+#     print("Test is now loaded")
+
+#     max_length = model.config.n_positions
+#     stride = 512
+#     seq_len = encodings.input_ids.size(1)
+
+#     nlls = []
+#     prev_end_loc = 0
+#     for begin_loc in tqdm(range(0, seq_len, stride)):
+#         end_loc = min(begin_loc + max_length, seq_len)
+#         trg_len = end_loc - prev_end_loc  # may be different from stride on last loop
+#         input_ids = encodings.input_ids[:, begin_loc:end_loc].to(device)
+#         target_ids = input_ids.clone()
+#         target_ids[:, :-trg_len] = -100
+
+#         with torch.no_grad():
+#             outputs = model(input_ids, labels=target_ids)
+
+#             # loss is calculated using CrossEntropyLoss which averages over valid labels
+#             # N.B. the model only calculates loss over trg_len - 1 labels, because it internally shifts the labels
+#             # to the left by 1.
+#             neg_log_likelihood = outputs.loss
+
+#         nlls.append(neg_log_likelihood)
+
+#         prev_end_loc = end_loc
+#         if end_loc == seq_len:
+#             break
+
+#     ppl = torch.exp(torch.stack(nlls).mean())
+#     return
```

Comment:
remove the unrelated code in this branch. You can move these code to a new branch for perplexity